pkg/config/config.go

```diff
@@ -1,201 +1,274 @@
-	return conf.store
+// Set sets a key in the store.
+func (c *Config) Set(key string, val interface{}) {
+	switch key {
+	case keyTimeout:
+		// go-toml requires int64
+		val = cast.ToInt64(val)
+	case keyPagination, keyReporting, keyPrompLogin:
+		val = cast.ToBool(val)
+	default:
+		val = cast.ToString(val)
+	}
+	c.tree.Set(key, val)
 }
 
-// SetPath sets a path for the Config.
-func (conf *Config) SetPath(path string) {
-	conf.store.path = path
+// Unset deletes a given key from the Config.
+func (c *Config) Unset(key string) {
+	keys := strings.Split(key, ".")
+
+	treeMap := c.tree.ToMap()
+	subMap := treeMap
+
+	// Extract a sub-map for each dotted key section.
+	for i, key := range keys {
+		if node, ok := subMap[key]; ok {
+			switch nodeType := node.(type) {
+			case map[string]interface{}:
+				subMap = nodeType
+			default:
+				// Abort if the section is not a map, unless it's the last section.
+				if i == len(keys)-1 {
+					// Remove the key from the parent map and swap trees.
+					delete(subMap, key)
+					c.tree, _ = toml.TreeFromMap(treeMap)
+				} else {
+					return
+				}
+			}
+		}
+	}
 }
 
-// Save unmarshals the Config into its associated Store and persists it to the disk.
-func (conf *Config) Save() error {
-	Marshal(conf, conf.store)
-	return conf.store.Persist()
+// Persist flushes the in-memory TOML tree representation to the path associated to the Config.
+func (c *Config) Persist() error {
+	if c.path == "" {
+		return ErrNoConfigPath
+	}
+	var buf bytes.Buffer
+	if _, err := c.tree.WriteTo(&buf); err != nil {
+		return err
+	}
+	return afero.WriteFile(c.fs, c.path, buf.Bytes(), 0600)
 }
 
-// TLS holds the configuration for TLS clients.
-type TLS struct {
-	// Insecure specifies if server certificates should be accepted without verification.
-	//
-	// Skipping verification against the system's CA bundle or a cluster-specific CA is highly discouraged
-	// and should only be done during testing/development.
-	Insecure bool
-
-	// Path to the root CA bundle.
-	RootCAsPath string
+// Keys returns all the keys in the Config.
+func (c *Config) Keys() []string {
+	var keys []string
+	for key, envVar := range c.envWhitelist {
+		if _, ok := c.envLookup(envVar); ok {
+			keys = append(keys, key)
+		}
+	}
 
-	// A pool of root CAs to verify server certificates against.
-	RootCAs *x509.CertPool
-}
+	searchKeys(c.tree, &keys, []string{})
+	keysLen := len(keys)
+	if keysLen < 2 {
+		// Less than 2 keys, no need to sort or remove duplicates.
+		return keys
+	}
 
-// New instanciates a configuration with all fields set to their zero values.
-func New() Config {
-	return Config{
-		store:       NewStore(StoreOpts{}),
-		dirtyFields: make(map[interface{}]bool),
+	sort.Strings(keys)
+
+	// Remove duplicates, this happens when there is both an env var and a TOML key.
+	// Keys are already sorted so this is done by comparing each key with the next one
+	// and removing the latter if it's the same.
+	for i := 1; i < keysLen; i++ {
```

Comment:
If we are sure that `keys` contains only two identical values at max and that they are sorted, the loop can be simplified:
```
// a = ["a", "a", "b", "c", "c"]
for i := 0; i < len(a) - 1; i++ {
	if a[i] == a[i+1] {
		a = append(a[:i], a[i+1:]...)
	}
}
// a = ["a", "b", "c"]
```
If they can be more than 2 times the same value, a `i--` has to be added in the loop.

Reply:
Good catch, indeed there cannot be more than 2 duplicates (one from env var and one from the config file), I've removed the `i--`.